File: orchestrator/workflow.py
"""High-level orchestration workflows used by CLI commands."""

from __future__ import annotations

import copy
from pathlib import Path
from typing import Any

from .compose_ops import compose_file_path, compose_running, generate_compose, run_compose_action
from .context import build_instance_context
from .errors import ValidationError
from .inventory import (
    find_instance,
    get_instances,
    inventory_path,
    load_inventory_file,
    save_inventory_file,
    validate_inventory,
)
from .policy import effective_policy_summary, validate_policies
from .render import render_instance_config
from .revisions import create_revision, list_revisions, load_revision_instance
from .utils import run_command, write_json


def load_and_validate(inv_file: str | None) -> tuple[Path, dict[str, Any]]:
    inv_path = inventory_path(inv_file)
    inventory = load_inventory_file(inv_path)
    validate_inventory(inventory, inv_path)
    validate_policies(inventory, get_instances(inventory))
    return inv_path, inventory


def validate_only(inv_file: str | None) -> tuple[Path, dict[str, Any]]:
    return load_and_validate(inv_file)


def render_instance(inv_file: str | None, instance_id: str, dry_run: bool = False) -> dict[str, Any]:
    inv_path, inventory = load_and_validate(inv_file)
    instance = find_instance(inventory, instance_id)
    context, rendered, generated_path = render_instance_config(instance, inv_path, dry_run=dry_run)

    if dry_run:
        return {
            "instance": instance_id,
            "dry_run": True,
            "generated_path": str(generated_path),
            "summary": {
                "agents": len((rendered.get("agents") or {}).get("list") or []),
                "bindings": len(rendered.get("bindings") or []),
                "channels": len(rendered.get("channels") or {}),
            },
        }

    return {
        "instance": instance_id,
        "dry_run": False,
        "generated_path": str(generated_path),
        "runtime_config_path": str(context.config_dir / "openclaw.json5"),
    }


def generate_compose_for_instance(inv_file: str | None, instance_id: str) -> dict[str, Any]:
    inv_path, inventory = load_and_validate(inv_file)
    instance = find_instance(inventory, instance_id)
    context, rendered, generated_path = render_instance_config(instance, inv_path, dry_run=False)
    compose = generate_compose(instance, context, generated_path)

    # Keep rendered config copy in generated path for troubleshooting.
    write_json(context.generated_dir / "openclaw.resolved.json", rendered)

    return {
        "instance": instance_id,
        "compose_path": str(compose),
    }


def run_compose(inv_file: str | None, instance_id: str, action: str) -> dict[str, Any]:
    inv_path, inventory = load_and_validate(inv_file)
    instance = find_instance(inventory, instance_id)
    context = build_instance_context(instance, inv_path)

    if action in {"up", "restart", "pull"}:
        render_instance_config(instance, inv_path, dry_run=False)
        generate_compose(instance, context, context.config_dir / "openclaw.json5")

    output = run_compose_action(context, action)
    return {
        "instance": instance_id,
        "action": action,
        "output": output,
    }


def preflight_instance(inv_file: str | None, instance_id: str) -> dict[str, Any]:
    inv_path, inventory = load_and_validate(inv_file)
    instance = find_instance(inventory, instance_id)

    docker_v = run_command(["docker", "--version"]).stdout.strip()
    compose_v = run_command(["docker", "compose", "version"]).stdout.strip()

    context, rendered, generated_path = render_instance_config(instance, inv_path, dry_run=False)
    compose_path = generate_compose(instance, context, generated_path)

    # Keep effective policies for operational inspection.
    policies = {
        "instance": effective_policy_summary(inventory, instance, None).policy,
        "agents": {},
    }
    for agent in instance.get("agents") or []:
        agent_id = agent.get("id")
        if isinstance(agent_id, str):
            policies["agents"][agent_id] = effective_policy_summary(
                inventory, instance, agent
            ).policy

    write_json(context.generated_dir / "effective-policy.json", policies)
    write_json(context.generated_dir / "render-summary.json", rendered)

    return {
        "instance": instance_id,
        "docker": docker_v,
        "compose": compose_v,
        "generated_config": str(generated_path),
        "generated_compose": str(compose_path),
        "effective_policy": str(context.generated_dir / "effective-policy.json"),
    }


def health_instance(inv_file: str | None, instance_id: str) -> dict[str, Any]:
    inv_path, inventory = load_and_validate(inv_file)
    instance = find_instance(inventory, instance_id)
    context = build_instance_context(instance, inv_path)

    running = compose_running(context)
    ps_text = run_compose_action(context, "ps") if compose_file_path(context).exists() else ""

    return {
        "instance": instance_id,
        "status": "running" if running else "degraded",
        "compose": str(compose_file_path(context)),
        "ps": ps_text,
    }


def update_instance(
    inv_file: str | None,
    instance_id: str,
    image_tag: str | None,
) -> dict[str, Any]:
    inv_path, inventory = load_and_validate(inv_file)
    instance = find_instance(inventory, instance_id)

    # Snapshot before mutating runtime config/image.
    context, rendered, generated_path = render_instance_config(instance, inv_path, dry_run=False)
    compose_path = generate_compose(instance, context, generated_path)
    revision = create_revision(inv_path, instance, rendered, compose_path)

    if image_tag:
        docker_cfg = (instance.setdefault("openclaw", {}).setdefault("docker", {}))
        image = str(docker_cfg.get("image") or "ghcr.io/openclaw/openclaw:latest")
        base = image.split(":", 1)[0]
        docker_cfg["image"] = f"{base}:{image_tag}"
        save_inventory_file(inv_path, inventory)

    # Re-run full validation post-mutation.
    validate_inventory(inventory, inv_path)
    validate_policies(inventory, get_instances(inventory))

    context, _, generated_path = render_instance_config(instance, inv_path, dry_run=False)
    generate_compose(instance, context, generated_path)
    pull_out = run_compose_action(context, "pull")
    up_out = run_compose_action(context, "up")

    status = "running" if compose_running(context) else "degraded"
    return {
        "instance": instance_id,
        "revision": revision,
        "status": status,
        "pull": pull_out,
        "up": up_out,
    }


def rollback_instance(inv_file: str | None, instance_id: str, revision: str) -> dict[str, Any]:
    inv_path = inventory_path(inv_file)
    inventory = load_inventory_file(inv_path)

    snap = load_revision_instance(instance_id, revision)

    instances = inventory.get("instances") or []
    replaced = False
    if isinstance(instances, list):
        for idx, instance in enumerate(instances):
            if isinstance(instance, dict) and instance.get("id") == instance_id:
                instances[idx] = copy.deepcopy(snap)
                replaced = True
                break

    if not replaced:
        raise ValidationError(f"instance not found in current inventory: {instance_id}")

    save_inventory_file(inv_path, inventory)
    validate_inventory(inventory, inv_path)
    validate_policies(inventory, get_instances(inventory))

    instance = find_instance(inventory, instance_id)
    context, _, generated_path = render_instance_config(instance, inv_path, dry_run=False)
    generate_compose(instance, context, generated_path)
    up_out = run_compose_action(context, "up")

    status = "running" if compose_running(context) else "degraded"
    return {
        "instance": instance_id,
        "revision": revision,
        "status": status,
        "up": up_out,
    }


def revisions_for_instance(instance_id: str) -> list[str]:
    return list_revisions(instance_id)

File: src/workflow.ts
import { existsSync } from 'node:fs';
import { composeFilePath, composeRunning, generateCompose, runComposeAction } from './compose';
import { buildInstanceContext } from './context';
import { resolveDeploymentProvider } from './deployment';
import { ValidationError } from './errors';
import {
  currentKubernetesContext,
  generateKubernetesManifest,
  kubernetesManifestPath,
  kubernetesRunning,
  resolveKubernetesTarget,
  runKubernetesAction,
  runKubernetesGatewayCommand,
} from './kubernetes';
import {
  findInstance,
  getInstances,
  inventoryPath,
  loadInventoryFile,
  saveInventoryFile,
  validateInventory,
} from './inventory';
import { effectivePolicySummary, validatePolicies } from './policy';
import { renderInstanceConfig } from './render';
import { createRevision, listRevisions, loadRevisionInstance } from './revisions';
import { runCommand, writeJson } from './utils';

export function loadAndValidate(invFile?: string): {
  invPath: string;
  inventory: Record<string, unknown>;
} {
  const invPath = inventoryPath(invFile);
  const inventory = loadInventoryFile(invPath);
  validateInventory(inventory, invPath);
  validatePolicies(inventory, getInstances(inventory));
  return { invPath, inventory };
}

export function validateOnly(invFile?: string): { invPath: string; inventory: Record<string, unknown> } {
  return loadAndValidate(invFile);
}

export function deploymentTargetForInstance(
  invFile: string | undefined,
  instanceId: string,
): Record<string, unknown> {
  const { inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);
  const resolved = resolveDeploymentProvider(inventory);

  if (resolved.provider === 'docker') {
    return {
      instance: instanceId,
      provider: resolved.provider,
      provider_source: resolved.source,
    };
  }

  const target = resolveKubernetesTarget(inventory, instance);
  return {
    instance: instanceId,
    provider: resolved.provider,
    provider_source: resolved.source,
    kubernetes: {
      namespace: target.namespace,
      context: target.context || currentKubernetesContext(target),
      kubeconfig: target.kubeconfig,
      deployment: target.deploymentName,
      service: target.serviceName,
    },
  };
}

export function renderInstance(
  invFile: string | undefined,
  instanceId: string,
  dryRun = false,
): Record<string, unknown> {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);

  const { context, rendered, generatedPath } = renderInstanceConfig(instance, invPath, dryRun);

  if (dryRun) {
    const agents = ((rendered.agents as Record<string, unknown>)?.list as unknown[]) ?? [];
    const bindings = (rendered.bindings as unknown[]) ?? [];
    const channels = Object.keys((rendered.channels as Record<string, unknown>) ?? {}).length;

    return {
      instance: instanceId,
      dry_run: true,
      generated_path: generatedPath,
      summary: {
        agents: agents.length,
        bindings: bindings.length,
        channels,
      },
    };
  }

  return {
    instance: instanceId,
    dry_run: false,
    generated_path: generatedPath,
    runtime_config_path: `${context.configDir}/openclaw.json5`,
  };
}

export function generateComposeForInstance(
  invFile: string | undefined,
  instanceId: string,
): Record<string, unknown> {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);

  const { context, rendered, generatedPath } = renderInstanceConfig(instance, invPath, false);
  const provider = resolveDeploymentProvider(inventory).provider;
  const runtimePath =
    provider === 'docker'
      ? generateCompose(instance, context, generatedPath)
      : generateKubernetesManifest(inventory, instance, context, generatedPath);

  writeJson(`${context.generatedDir}/openclaw.resolved.json`, rendered);

  return {
    instance: instanceId,
    provider,
    runtime_manifest: runtimePath,
  };
}

export function runCompose(
  invFile: string | undefined,
  instanceId: string,
  action: string,
): Record<string, unknown> {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);
  const provider = resolveDeploymentProvider(inventory).provider;
  const context = buildInstanceContext(instance, invPath);

  if (['up', 'restart', 'pull'].includes(action)) {
    renderInstanceConfig(instance, invPath, false);
    if (provider === 'docker') {
      generateCompose(instance, context, `${context.configDir}/openclaw.json5`);
    } else {
      generateKubernetesManifest(inventory, instance, context, `${context.configDir}/openclaw.json5`);
    }
  }

  const output =
    provider === 'docker'
      ? runComposeAction(context, action)
      : runKubernetesAction(inventory, instance, context, action);
  return {
    instance: instanceId,
    provider,
    action,
    output,
  };
}

export function pairingList(
  invFile: string | undefined,
  instanceId: string,
  channel: string,
  accountId: string,
  jsonOutput = false,
): Record<string, unknown> {
  const args = ['pairing', 'list', channel, '--account', accountId];
  if (jsonOutput) {
    args.push('--json');
  }

  const raw = runGatewayCommand(invFile, instanceId, args);
  return {
    instance: instanceId,
    channel,
    account: accountId,
    output: jsonOutput ? parseOutputJson(raw) : raw,
  };
}

export function pairingApprove(
  invFile: string | undefined,
  instanceId: string,
  channel: string,
  code: string,
  accountId: string,
): Record<string, unknown> {
  const raw = runGatewayCommand(invFile, instanceId, [
    'pairing',
    'approve',
    channel,
    code,
    '--account',
    accountId,
  ]);

  return {
    instance: instanceId,
    channel,
    account: accountId,
    output: raw,
  };
}

export function preflightInstance(
  invFile: string | undefined,
  instanceId: string,
): Record<string, unknown> {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);
  const provider = resolveDeploymentProvider(inventory).provider;

  const { context, rendered, generatedPath } = renderInstanceConfig(instance, invPath, false);
  const runtimePath =
    provider === 'docker'
      ? generateCompose(instance, context, generatedPath)
      : generateKubernetesManifest(inventory, instance, context, generatedPath);

  const policies: Record<string, unknown> = {
    instance: effectivePolicySummary(inventory, instance).policy,
    agents: {},
  };

  const agents = Array.isArray(instance.agents) ? instance.agents : [];
  for (const rawAgent of agents) {
    if (!isRecord(rawAgent)) {
      continue;
    }
    const agentId = typeof rawAgent.id === 'string' ? rawAgent.id : '';
    if (!agentId) {
      continue;
    }
    (policies.agents as Record<string, unknown>)[agentId] = effectivePolicySummary(
      inventory,
      instance,
      rawAgent,
    ).policy;
  }

  writeJson(`${context.generatedDir}/effective-policy.json`, policies);
  writeJson(`${context.generatedDir}/render-summary.json`, rendered);

  if (provider === 'docker') {
    const dockerVersion = runCommand(['docker', '--version']).stdout.trim();
    const composeVersion = runCommand(['docker', 'compose', 'version']).stdout.trim();

    return {
      instance: instanceId,
      provider,
      docker: dockerVersion,
      compose: composeVersion,
      generated_config: generatedPath,
      runtime_manifest: runtimePath,
      effective_policy: `${context.generatedDir}/effective-policy.json`,
    };
  }

  const target = resolveKubernetesTarget(inventory, instance);
  const kubectlVersion = runCommand(['kubectl', 'version', '--client']).stdout.trim();
  const kubeContext = target.context || currentKubernetesContext(target);

  return {
    instance: instanceId,
    provider,
    kubectl: kubectlVersion,
    kube_context: kubeContext,
    kube_namespace: target.namespace,
    generated_config: generatedPath,
    runtime_manifest: runtimePath,
    effective_policy: `${context.generatedDir}/effective-policy.json`,
  };
}

export function healthInstance(invFile: string | undefined, instanceId: string): Record<string, unknown> {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);
  const provider = resolveDeploymentProvider(inventory).provider;
  const context = buildInstanceContext(instance, invPath);

  if (provider === 'docker') {
    const running = composeRunning(context);
    const composePath = composeFilePath(context);
    const ps = existsSync(composePath) ? runComposeAction(context, 'ps') : '';

    return {
      instance: instanceId,
      provider,
      status: running ? 'running' : 'degraded',
      runtime_manifest: composePath,
      ps,
    };
  }

  const running = kubernetesRunning(inventory, instance);
  const manifestPath = kubernetesManifestPath(context);
  const ps = runKubernetesAction(inventory, instance, context, 'ps');

  return {
    instance: instanceId,
    provider,
    status: running ? 'running' : 'degraded',
    runtime_manifest: manifestPath,
    ps,
  };
}

export function updateInstance(
  invFile: string | undefined,
  instanceId: string,
  imageTag?: string,
): Record<string, unknown> {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);
  const provider = resolveDeploymentProvider(inventory).provider;

  const rendered = renderInstanceConfig(instance, invPath, false);
  const runtimePath =
    provider === 'docker'
      ? generateCompose(instance, rendered.context, rendered.generatedPath)
      : generateKubernetesManifest(inventory, instance, rendered.context, rendered.generatedPath);
  const revision = createRevision(invPath, instance, rendered.rendered, runtimePath);

  if (imageTag) {
    updateImageTag(instance, provider, imageTag);
    saveInventoryFile(invPath, inventory);
  }

  validateInventory(inventory, invPath);
  validatePolicies(inventory, getInstances(inventory));

  const rerender = renderInstanceConfig(instance, invPath, false);
  if (provider === 'docker') {
    generateCompose(instance, rerender.context, rerender.generatedPath);
    const pullOutput = runComposeAction(rerender.context, 'pull');
    const upOutput = runComposeAction(rerender.context, 'up');
    return {
      instance: instanceId,
      provider,
      revision,
      status: composeRunning(rerender.context) ? 'running' : 'degraded',
      pull: pullOutput,
      up: upOutput,
    };
  }

  generateKubernetesManifest(inventory, instance, rerender.context, rerender.generatedPath);
  const upOutput = runKubernetesAction(inventory, instance, rerender.context, 'up');
  const pullOutput = runKubernetesAction(inventory, instance, rerender.context, 'pull');

  return {
    instance: instanceId,
    provider,
    revision,
    status: kubernetesRunning(inventory, instance) ? 'running' : 'degraded',
    pull: pullOutput,
    up: upOutput,
  };
}

export function rollbackInstance(
  invFile: string | undefined,
  instanceId: string,
  revision: string,
): Record<string, unknown> {
  const invPath = inventoryPath(invFile);
  const inventory = loadInventoryFile(invPath);
  const snapshot = loadRevisionInstance(instanceId, revision);

  if (!Array.isArray(inventory.instances)) {
    throw new ValidationError('instances must be a list');
  }

  let replaced = false;
  for (let i = 0; i < inventory.instances.length; i += 1) {
    const raw = inventory.instances[i];
    if (typeof raw === 'object' && raw !== null && !Array.isArray(raw)) {
      const item = raw as Record<string, unknown>;
      if (item.id === instanceId) {
        inventory.instances[i] = snapshot;
        replaced = true;
        break;
      }
    }
  }

  if (!replaced) {
    throw new ValidationError(`instance not found in current inventory: ${instanceId}`);
  }

  saveInventoryFile(invPath, inventory);
  validateInventory(inventory, invPath);
  validatePolicies(inventory, getInstances(inventory));

  const instance = findInstance(inventory, instanceId);
  const provider = resolveDeploymentProvider(inventory).provider;
  const render = renderInstanceConfig(instance, invPath, false);

  if (provider === 'docker') {
    generateCompose(instance, render.context, render.generatedPath);
    const upOutput = runComposeAction(render.context, 'up');
    return {
      instance: instanceId,
      provider,
      revision,
      status: composeRunning(render.context) ? 'running' : 'degraded',
      up: upOutput,
    };
  }

  generateKubernetesManifest(inventory, instance, render.context, render.generatedPath);
  const upOutput = runKubernetesAction(inventory, instance, render.context, 'up');
  return {
    instance: instanceId,
    provider,
    revision,
    status: kubernetesRunning(inventory, instance) ? 'running' : 'degraded',
    up: upOutput,
  };
}

export function revisionsForInstance(instanceId: string): string[] {
  return listRevisions(instanceId);
}

function runGatewayCommand(
  invFile: string | undefined,
  instanceId: string,
  commandArgs: string[],
): string {
  const { invPath, inventory } = loadAndValidate(invFile);
  const instance = findInstance(inventory, instanceId);
  const provider = resolveDeploymentProvider(inventory).provider;
  if (provider === 'kubernetes') {
    return runKubernetesGatewayCommand(inventory, instance, commandArgs);
  }

  const context = buildInstanceContext(instance, invPath);
  const composePath = composeFilePath(context);

  if (!existsSync(composePath)) {
    throw new ValidationError(
      `compose file not generated: ${composePath}. Run 'oco compose up --instance ${instanceId}' first.`,
    );
  }

  const service = resolveComposeService(instance);
  const args = [
    'docker',
    'compose',
    '-f',
    composePath,
    'exec',
    '-T',
    service,
    'node',
    '/app/openclaw.mjs',
    ...commandArgs,
  ];

  const result = runCommand(args);
  return (result.stdout || result.stderr).trim();
}

function resolveComposeService(instance: Record<string, unknown>): string {
  const openclaw = isRecord(instance.openclaw) ? instance.openclaw : {};
  const docker = isRecord(openclaw.docker) ? openclaw.docker : {};

  if (typeof docker.service_name === 'string' && docker.service_name.trim()) {
    return docker.service_name.trim();
  }

  return 'gateway';
}

function parseOutputJson(raw: string): unknown {
  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}

function updateImageTag(instance: Record<string, unknown>, provider: 'docker' | 'kubernetes', imageTag: string): void {
  if (!isRecord(instance.openclaw)) {
    instance.openclaw = {};
  }
  const openclaw = instance.openclaw as Record<string, unknown>;

  if (provider === 'docker') {
    if (!isRecord(openclaw.docker)) {
      openclaw.docker = {};
    }
    const docker = openclaw.docker as Record<string, unknown>;
    const image = typeof docker.image === 'string' ? docker.image : 'ghcr.io/openclaw/openclaw:latest';
    const base = image.includes(':') ? image.split(':', 1)[0] : image;
    docker.image = `${base}:${imageTag}`;
    return;
  }

  if (!isRecord(openclaw.kubernetes)) {
    openclaw.kubernetes = {};
  }
  const kubernetes = openclaw.kubernetes as Record<string, unknown>;
  const docker = isRecord(openclaw.docker) ? (openclaw.docker as Record<string, unknown>) : {};
  const image =
    (typeof kubernetes.image === 'string' && kubernetes.image.trim()) ||
    (typeof docker.image === 'string' && docker.image.trim()) ||
    'ghcr.io/openclaw/openclaw:latest';
  const base = image.includes(':') ? image.split(':', 1)[0] : image;
  kubernetes.image = `${base}:${imageTag}`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
